pkg/compiler/validators/bindings.go

```diff
@@ -125,6 +125,12 @@ func validateBinding(w c.WorkflowBuilder, nodeID c.NodeID, nodeParam string, bin
 					}
 				}
 
+				// Skip the validation if the promise has attribute paths
+				// because we don't know the type of the resolved attribute
+				if len(val.Promise.AttrPath) > 0 {
```

Comment:
we should be able to know the attribute type if the promise type is list or dict, right?

Reply:
We should be able to use the [AreTypesCastable function](https://github.com/flyteorg/flytepropeller/blob/6b1d11c736c4a9d8b8fa48264de0d3dc5853d494/pkg/compiler/validators/typing.go#L358) here right? It should handle `Union` as either a source or destination type. IMO this feature needs type validation, it's something Flyte is very opinionated about and seems very sloppy to leave out.